File: WordPress/WordPressTest/NotificationTests.m
#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import "ContextManager-Internals.h"
#import "TestContextManager.h"
#import "Notification.h"


@interface NotificationTests : XCTestCase
@property (nonatomic, strong) TestContextManager *contextManager;
@end

@implementation NotificationTests

- (void)setUp
{
    [super setUp];
    self.contextManager = [[TestContextManager alloc] init];
}

- (void)tearDown
{
    [super tearDown];
    
    // Note: We'll force TestContextManager override reset, since, for (unknown reasons) the TestContextManager
    // might be retained more than expected, and it may break other core data based tests.
    [ContextManager overrideSharedInstance:nil];
}

- (void)testBadgeNotificationHasBadgeFlagSetToTrue
{
    Notification *note = [self loadBadgeNotification];
    XCTAssertTrue(note.isBadge, @"Badge flag should be true");
}

- (void)testBadgeNotificationHasRegularFieldsSet
{
    Notification *note = [self loadBadgeNotification];
    XCTAssertNotNil(note.type, @"Missing Notification Type");
    XCTAssertNotNil(note.noticon, @"Missing Noticon");
    XCTAssertNotNil(note.timestampAsDate, @"Timestamp could not be parsed");
    XCTAssertNotNil(note.icon, @"Missing Noticon");
    XCTAssertNotNil(note.url, @"Missing Resource URL");
}

- (void)testBadgeNotificationContainsOneSubjectBlock
{
    Notification *note = [self loadBadgeNotification];
    XCTAssertNotNil(note.subjectBlock, @"Missing subjectBlock");
    XCTAssertNotNil(note.subjectBlock.text, @"Subject Block has no text!");
}

- (void)testBadgeNotificationContainsOneImageBlockGroup
{
    Notification *note = [self loadBadgeNotification];
    NotificationBlockGroup *group = [note blockGroupOfType:NoteBlockGroupTypeImage];
    XCTAssertNotNil(group, @"Missing Image Block Group");
    
    NotificationBlock *imageBlock = group.blocks.firstObject;
    XCTAssertNotNil(imageBlock, @"Missing Image Block");
    
    NotificationMedia *media = imageBlock.media.firstObject;
    XCTAssertNotNil(media, @"Missing Media");
    XCTAssertNotNil(media.mediaURL, @"Missing mediaURL");
}

- (void)testLikeNotificationContainsOneSubjectBlock
{
    Notification *note = [self loadLikeNotification];
    XCTAssertNotNil(note.subjectBlock, @"Missing subjectBlock");
    XCTAssertNotNil(note.subjectBlock.text, @"Subject Block has no text!");
}

- (void)testLikeNotificationContainsHeader
{
    Notification *note = [self loadLikeNotification];
    NotificationBlockGroup *header = note.headerBlockGroup;
    XCTAssertNotNil(header, @"Missing headerBlockGroup");

    NotificationBlock *gravatarBlock = [header blockOfType:NoteBlockTypeImage];
    XCTAssertNotNil(gravatarBlock.text, @"Missing Gravatar Text");
    
    NotificationMedia *media = gravatarBlock.media.firstObject;
    XCTAssertNotNil(media.mediaURL, @"Missing Gravatar URL");
    
    NotificationBlock *snippetBlock = [header blockOfType:NoteBlockTypeText];
    XCTAssertNotNil(snippetBlock.text, @"Missing Snippet Block");
}

- (void)testLikeNotificationContainsUserBlocksInTheBody
{
    Notification *note = [self loadLikeNotification];
    for (NotificationBlockGroup *group in note.bodyBlockGroups) {
        XCTAssertTrue(group.type == NoteBlockGroupTypeUser);
    }
}

- (void)testLikeNotificationContainsPostAndSiteID
{
    Notification *note = [self loadLikeNotification];
    XCTAssertNotNil(note.metaSiteID, @"Missing SiteID");
    XCTAssertNotNil(note.metaPostID, @"Missing PostID");
}

- (void)testFollowerNotificationHasFollowFlagSetToTrue
{
    Notification *note = [self loadFollowerNotification];
    XCTAssertTrue(note.isFollow, @"Follow flag should be true");
}

- (void)testFollowerNotificationContainsOneSubjectBlock
{
    Notification *note = [self loadFollowerNotification];
    XCTAssertNotNil(note.subjectBlock, @"Missing subjectBlock");
    XCTAssertNotNil(note.subjectBlock.text, @"Subject Block has no text!");
}

- (void)testFollowerNotificationContainsSiteID
{
    Notification *note = [self loadFollowerNotification];
    XCTAssertNotNil(note.metaSiteID, @"Missing siteID");
}

- (void)testFollowerNotificationContainsUserAndFooterBlocksInTheBody
{
    Notification *note = [self loadFollowerNotification];
    
    // Note: Account for 'View All Followers'
    for (NotificationBlockGroup *group in note.bodyBlockGroups) {
        XCTAssertTrue(group.type == NoteBlockGroupTypeUser || group.type == NoteBlockGroupTypeFooter);
    }
}

- (void)testFollowerNotificationContainsFooterBlockWithFollowRangeAtTheEnd
{
    Notification *note = [self loadFollowerNotification];
    
    NotificationBlockGroup *lastGroup = note.bodyBlockGroups.lastObject;
    XCTAssertTrue(lastGroup.type == NoteBlockGroupTypeFooter);
    
    NotificationBlock *block = [lastGroup.blocks firstObject];
    XCTAssertNotNil(block.text, @"Missing block text");
    
    NotificationRange *range = block.ranges.firstObject;
    XCTAssertNotNil(range, @"Missing range");
    XCTAssertEqualObjects(range.type, @"follow", @"Missing follow range");
}

- (void)testCommentNotificationHasCommentFlagSetToTrue
{
    Notification *note = [self loadCommentNotification];
    XCTAssertTrue(note.isComment, @"Comment flag should be true");
}

- (void)testCommentNotificationContainsSubjectWithSnippet
{
    Notification *note = [self loadCommentNotification];
    XCTAssertNotNil(note.subjectBlock.text, @"Subject Block has no text!");
    XCTAssertNotNil(note.snippetBlock.text, @"Subject Block has no text!");
}

- (void)testCommentNotificationContainsHeader
{
    Notification *note = [self loadCommentNotification];
    NotificationBlockGroup *header = note.headerBlockGroup;
    XCTAssertNotNil(header, @"Missing headerBlockGroup");
    
    NotificationBlock *gravatarBlock = [header blockOfType:NoteBlockTypeImage];
    XCTAssertNotNil(gravatarBlock.text, @"Missing Gravatar Text");
    
    NotificationMedia *media = gravatarBlock.media.firstObject;
    XCTAssertNotNil(media.mediaURL, @"Missing Gravatar URL");
    
    NotificationBlock *snippetBlock = [header blockOfType:NoteBlockTypeText];
    XCTAssertNotNil(snippetBlock.text, @"Missing Snippet Block");
}

- (void)testCommentNotificationContainsCommentAndSiteID
{
    Notification *note = [self loadCommentNotification];
    XCTAssertNotNil(note.metaSiteID, @"Missing siteID");
    XCTAssertNotNil(note.metaCommentID, @"Missing commentID");
}

- (void)testFindingNotificationRangeSearchingByReplyCommentID
{
    Notification *note = [self loadCommentNotification];
    XCTAssertNotNil(note.metaReplyID, @"Missing Reply ID");
    
    NotificationBlockGroup *footerGroup = [note blockGroupOfType:NoteBlockGroupTypeFooter];
    NotificationBlock *textBlock = [footerGroup blockOfType:NoteBlockGroupTypeText];
    XCTAssertNotNil(textBlock, @"Missing View Reply Text Block");
    
    NotificationRange *range = [textBlock notificationRangeWithCommentId:note.metaReplyID];
    XCTAssertNotNil(range, @"Missing Reply Range");
}


#pragma mark - Helpers

- (NSString *)entityName
{
    return NSStringFromClass([Notification class]);
}

- (Notification *)loadBadgeNotification
{
    return (Notification *)[self.contextManager loadEntityNamed:self.entityName
                                             withContentsOfFile:@"notifications-badge.json"];
}

- (Notification *)loadLikeNotification
{
    return (Notification *)[self.contextManager loadEntityNamed:self.entityName
                                             withContentsOfFile:@"notifications-like.json"];
}

- (Notification *)loadFollowerNotification
{
    return (Notification *)[self.contextManager loadEntityNamed:self.entityName
                                             withContentsOfFile:@"notifications-new-follower.json"];
}

- (Notification *)loadCommentNotification
{
    return (Notification *)[self.contextManager loadEntityNamed:self.entityName
                                             withContentsOfFile:@"notifications-replied-comment.json"];
}

@end

File: WordPress/WordPressTest/NotificationTests.swift
import Foundation
import XCTest
@testable import WordPress



/// Notifications Tests
///
class NotificationTests: CoreDataTestCase {

    private var utility: NotificationUtility!

    override func setUp() {
        utility = NotificationUtility(coreDataStack: contextManager)
    }

    override func tearDown() {
        utility = nil
    }

    func testBadgeNotificationHasBadgeFlagSetToTrue() throws {
        let note = try loadBadgeNotification()
        XCTAssertTrue(note.isBadge)
    }

    func testBadgeNotificationHasRegularFieldsSet() throws {
        let note = try loadBadgeNotification()
        XCTAssertNotNil(note.type)
        XCTAssertNotNil(note.noticon)
        XCTAssertNotNil(note.iconURL)
        XCTAssertNotNil(note.resourceURL)
        XCTAssertNotNil(note.timestampAsDate)
    }

    func testBadgeNotificationProperlyLoadsItsSubjectContent() throws {
        let note = try utility.loadBadgeNotification()

        XCTAssert(note.subjectContentGroup?.blocks.count == 1)
        XCTAssertNotNil(note.subjectContentGroup?.blocks.first)
        XCTAssertNotNil(note.renderSubject())
    }

    func testBadgeNotificationContainsOneImageContentGroup() throws {
        let note = try utility.loadBadgeNotification()
        let group = note.contentGroup(ofKind: .image)
        XCTAssertNotNil(group)

        let imageBlock = group?.blocks.first as? FormattableMediaContent
        XCTAssertNotNil(imageBlock)

        let media = imageBlock?.media.first
        XCTAssertNotNil(media)
        XCTAssertNotNil(media?.mediaURL)
    }

    func testLikeNotificationReturnsTheProperKindValue() throws {
        let note = try loadLikeNotification()
        XCTAssert(note.kind == .like)
    }

    func testLikeNotificationContainsHeaderContent() throws {
        let note = try loadLikeNotification()
        let header = note.headerContentGroup
        XCTAssertNotNil(header)

        let gravatarBlock: NotificationTextContent? = header?.blockOfKind(.image)
        XCTAssertNotNil(gravatarBlock?.text)

        let media = gravatarBlock?.media.first
        XCTAssertNotNil(media?.mediaURL)

        let snippetBlock: NotificationTextContent? = header?.blockOfKind(.text)
        XCTAssertNotNil(snippetBlock?.text)
    }


    func testLikeNotificationContainsUserContentGroupsInTheBody() throws {
        let note = try utility.loadLikeNotification()
        for group in note.bodyContentGroups {
            XCTAssertTrue(group.kind == .user)
        }
    }

    func testLikeNotificationContainsPostAndSiteID() throws {
        let note = try loadLikeNotification()
        XCTAssertNotNil(note.metaSiteID)
        XCTAssertNotNil(note.metaPostID)
    }

    func testFollowerNotificationReturnsTheProperKindValue() throws {
        let note = try loadFollowerNotification()
        XCTAssert(note.kind == .follow)
    }

    func testFollowerNotificationHasFollowFlagSetToTrue() throws {
        let note = try loadFollowerNotification()
        XCTAssertTrue(note.kind == .follow)
    }

    func testFollowerNotificationContainsOneSubjectContent() throws {
        let note = try loadFollowerNotification()

        let content = note.subjectContentGroup?.blocks.first
        XCTAssertNotNil(content)
        XCTAssertNotNil(content?.text)
    }

    func testFollowerNotificationContainsSiteID() throws {
        let note = try loadFollowerNotification()
        XCTAssertNotNil(note.metaSiteID)
    }

    func testFollowerNotificationContainsUserAndFooterGroupsInTheBody() throws {
        let note = try utility.loadFollowerNotification()

        // Note: Account for 'View All Followers'
        for group in note.bodyContentGroups {
            XCTAssertTrue(group.kind == .user || group.kind == .footer)
        }
    }

    func testFollowerNotificationContainsFooterContentWithFollowRangeAtTheEnd() throws {
        let note = try loadFollowerNotification()

        let lastGroup = note.bodyContentGroups.last
        XCTAssertNotNil(lastGroup)
        XCTAssertTrue(lastGroup!.kind == .footer)

        let block = lastGroup?.blocks.first
        XCTAssertNotNil(block)
        XCTAssertNotNil(block?.text)
        XCTAssertNotNil(block?.ranges)

        let range = block?.ranges.last
        XCTAssertNotNil(range)
        XCTAssert(range?.kind == .follow)
    }

    func testCommentNotificationReturnsTheProperKindValue() throws {
        let note = try loadCommentNotification()
        XCTAssert(note.kind == .comment)
    }

    func testCommentNotificationHasCommentFlagSetToTrue() throws {
        let note = try loadCommentNotification()
        XCTAssertTrue(note.kind == .comment)
    }

    func testCommentNotificationRendersSubjectWithSnippet() throws {
        let note = try loadCommentNotification()

        XCTAssertNotNil(note.renderSubject())
        XCTAssertNotNil(note.renderSnippet())
    }

    func testCommentNotificationContainsHeaderContent() throws {
        let note = try loadCommentNotification()

        let header = note.headerContentGroup
        XCTAssertNotNil(header)

        let gravatarBlock: NotificationTextContent? = header?.blockOfKind(.image)
        XCTAssertNotNil(gravatarBlock)
        XCTAssertNotNil(gravatarBlock?.text)

        let media = gravatarBlock!.media.first
        XCTAssertNotNil(media)
        XCTAssertNotNil(media!.mediaURL)

        let snippetBlock: NotificationTextContent? = header?.blockOfKind(.text)
        XCTAssertNotNil(snippetBlock)
        XCTAssertNotNil(snippetBlock?.text)
    }

    func testCommentNotificationContainsCommentAndSiteID() throws {
        let note = try loadCommentNotification()
        XCTAssertNotNil(note.metaSiteID)
        XCTAssertNotNil(note.metaCommentID)
    }

    func testCommentNotificationProperlyChecksIfItWasRepliedTo() throws {
        let note = try loadCommentNotification()
        XCTAssert(note.isRepliedComment)
    }

    func testCommentNotificationIsUnapproved() throws {
        let note = try utility.loadUnapprovedCommentNotification()
        XCTAssertTrue(note.isUnapprovedComment)
    }

    func testCommentNotificationIsApproved() throws {
        let note = try utility.loadCommentNotification()
        XCTAssertFalse(note.isUnapprovedComment)
    }


    func testFooterContentIsIdentifiedAndCreated() throws {
        let note = try loadCommentNotification()
        let footerBlock: FooterTextContent? = note.contentGroup(ofKind: .footer)?.blockOfKind(.text)

        XCTAssertNotNil(footerBlock)
    }

    func testFindingContentRangeSearchingByURL() throws {
        let note = try loadBadgeNotification()
        let targetURL = URL(string: "http://www.wordpress.com")!
        let range = note.contentRange(with: targetURL)

        XCTAssertNotNil(range)
    }

    func testPingbackNotificationIsPingback() throws {
        let notification = try utility.loadPingbackNotification()
        XCTAssertTrue(notification.isPingback)
    }

    func testPingbackBodyContainsFooter() throws {
        let notification = try utility.loadPingbackNotification()
        let footer = notification.bodyContentGroups.filter { $0.kind == .footer }
        XCTAssertEqual(footer.count, 1)
    }

    func testHeaderAndBodyContentGroups() throws {
        let note = try utility.loadCommentNotification()
        let headerGroupsCount = note.headerContentGroup != nil ? 1 : 0
        let bodyGroupsCount = note.bodyContentGroups.count
        let totalGroupsCount = headerGroupsCount + bodyGroupsCount

        XCTAssertEqual(note.headerAndBodyContentGroups.count, totalGroupsCount)
    }

    func testNotificationCacheIsInvalidated() throws {
        let commentNotificationId = "44444"
        let _ = try utility.loadCommentNotification()
        contextManager.saveContextAndWait(mainContext)

        let fetchRequest = NSFetchRequest<WordPress.Notification>(entityName: WordPress.Notification.entityName())
        fetchRequest.predicate = NSPredicate(format: "notificationId == %@", commentNotificationId)

        let note = try XCTUnwrap(mainContext.fetch(fetchRequest).first)
        XCTAssertEqual(note.timestamp, "2015-03-10T18:57:37+00:00")
        XCTAssertEqual(note.timestampAsDate.timeIntervalSince1970, 1426013857)

        note.timestamp = "2015-03-10T18:57:38+00:00"
        XCTAssertEqual(note.timestampAsDate.timeIntervalSince1970, 1426013858)

        mainContext.reset()

        contextManager.performAndSave { context in
            let notification = (try? context.fetch(fetchRequest))?.first
            XCTAssertNotNil(notification)
            XCTAssertEqual(notification?.timestampAsDate.timeIntervalSince1970, 1426013857)

            XCTExpectFailure("""
                This assertion failure is problematic.
                When timestamp (or any other cached attribute) is changed, the cache still should be invalidated
                even though the notification is associated with a non-main context.
                """)
            note.timestamp = "2015-03-10T18:57:38+00:00"
            XCTAssertEqual(notification?.timestampAsDate.timeIntervalSince1970, 1426013858)
        }
    }

    // MARK: - Helpers

    func loadBadgeNotification() throws -> WordPress.Notification {
        return try utility.loadBadgeNotification()
    }

    func loadLikeNotification() throws -> WordPress.Notification {
        return try utility.loadLikeNotification()
    }

    func loadFollowerNotification() throws -> WordPress.Notification {
        return try utility.loadFollowerNotification()
    }

    func loadCommentNotification() throws -> WordPress.Notification {
        return try utility.loadCommentNotification()
    }
}
